File: packages/tools/src/write.ts
import { promises as fs } from "node:fs";
import { dirname } from "node:path";
import { type Tool, lineDiff, writePreview } from "@arterm/core";
import { requireString, resolveWithin } from "./paths.js";
import { invalidateSearchIndex } from "./search.js";

export const writeTool: Tool = {
  name: "write",
  description: "Create or overwrite a file with the given content.",
  permission: "ask",
  category: "edit",
  mutating: true,
  riskTier: "caution",
  parameters: {
    type: "object",
    properties: {
      path: { type: "string", description: "File path relative to the working directory." },
      content: { type: "string", description: "Full file content to write." },
    },
    required: ["path", "content"],
  },
  preview: (args) =>
    writePreview(String(args.path), typeof args.content === "string" ? args.content : ""),
  async execute(args, ctx) {
    const relPath = requireString(args, "path");
    const abs = resolveWithin(ctx.cwd, relPath);
    const content = requireString(args, "content");
    // Read the prior content (if any) so an overwrite renders as a real diff.
    const before = await fs.readFile(abs, "utf8").catch(() => "");
    await fs.mkdir(dirname(abs), { recursive: true });
    await fs.writeFile(abs, content, "utf8");
    invalidateSearchIndex(ctx.cwd);
    return {
      output: `Wrote ${content.length} bytes to ${relPath}`,
      diff: lineDiff(before, content),
      path: relPath,
    };
  },
};

File: rust/crates/arterm-tools/src/write.rs
use anyhow::Result;
use async_trait::async_trait;

use arterm_core::Permission;
use crate::Tool;

pub struct WriteTool;

#[async_trait]
impl Tool for WriteTool {
    fn name(&self) -> &str { "write" }
    fn description(&self) -> &str { "Write content to a file, creating or overwriting it." }
    fn parameters(&self) -> serde_json::Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "path": { "type": "string" },
                "content": { "type": "string" }
            },
            "required": ["path", "content"]
        })
    }
    fn permission(&self) -> Permission { Permission::Write }

    async fn execute(&self, args: serde_json::Value) -> Result<String> {
        let path = args["path"].as_str().ok_or_else(|| anyhow::anyhow!("missing 'path'"))?;
        let content = args["content"].as_str().ok_or_else(|| anyhow::anyhow!("missing 'content'"))?;
        tokio::fs::write(path, content).await
            .map_err(|e| anyhow::anyhow!("write {path}: {e}"))?;
        Ok(format!("wrote {} bytes to {path}", content.len()))
    }
}
